engineering: add configuration guidelines

smola opened the pull request:
We use environment variables for configuration whenever possible.
There are some exceptions when using Spark.

Signed-off-by: Santiago M. Mola <santi@mola.io>

juanjux commented on engineering/conventions.md:
```diff
@@ -53,4 +54,11 @@ Refer to the [licensing](licensing.md), [maintainers](maintainers.md) and [docum
 
-Each sourcerer has his own preferences on editors and IDEs. But if you do not know what to use, [Visual Studio Code](https://code.visualstudio.com/) is a popular choice for Go and Python, and [IntelliJ IDEA](https://www.jetbrains.com/idea/) is the most used for Java and Scala.
\ No newline at end of file
+Each sourcerer has his own preferences on editors and IDEs. But if you do not know what to use, [Visual Studio Code](https://code.visualstudio.com/) is a popular choice for Go and Python, and [IntelliJ IDEA](https://www.jetbrains.com/idea/) is the most used for Java and Scala.
```
PyCharm is pretty popular too with pythonists (I would say more than VS Code).

ajnavarro commented on engineering/conventions.md:
```diff
@@ -53,4 +54,11 @@ Refer to the [licensing](licensing.md), [maintainers](maintainers.md) and [docum
+
+## Configuration
+
+Your application should be [configured](https://12factor.net/config) using environment variables. External services (e.g. databases) URLs, storage paths, [ports to bind](https://12factor.net/port-binding) and logging format all belong to configuration.
```
So, if we have a lot of small configuration params, really common on databases to tweak them for specific use cases, are we going to put all of them using environment variables?

rporres replied:
Yes, we definitely should do that. Otherwise it is painful to pass config files to docker. In k8s you need to create a config map and then pass it to the pod to mount it, which is feasible, but it is a bit painful

smola replied:
@ajnavarro In that case it's a good idea to allow connection string style.
For example:
`DATABASE=postgres://user:pass@host:port/database?sslmode=verify-full`

smola replied:
Actually, not only in that case, it's a good idea most of the time. For example:
`BROKER=amqp://host:port` or `BROKER=memory://`.

mcuadros commented on engineering/conventions-go.md:
```diff
@@ -62,4 +62,10 @@ A blank line is usually added after the end of any control structure, too.
+## Configuration
+
+* Use environment variables for configuration.
+* You can use the [configurable](https://github.com/src-d/framework/tree/master/configurable) package  ([example](https://github.com/src-d/framework/blob/master/database/database.go)).
```
I don't know if propagate `framework` to new projects is a good, idea.

smola replied:
Sure. `configurable` package is so small we can probably go ahead and recommend [`envconfig`](https://github.com/kelseyhightower/envconfig)?

mcuadros commented on engineering/conventions.md:
```diff
@@ -53,4 +54,11 @@ Refer to the [licensing](licensing.md), [maintainers](maintainers.md) and [docum
+
+## Configuration
+
+Your application should be [configured](https://12factor.net/config) using environment variables. External services (e.g. databases) URLs, storage paths, [ports to bind](https://12factor.net/port-binding) and logging format all belong to configuration.
```
Will be nice include the naming conventions for the environment variables.

dpordomingo replied:
we (in apps) use a prefix for all our env vars,
example: in `src-d/code-annotation` we choosed `CAT` prefix, so we use `CAT_HOST` instead of `HOST`

smola commented:
Updated.

@mcuadros 

> I don't know if propagate framework to new projects is a good, idea.

Changed to recommend `envconfig` directly.

> Will be nice include the naming conventions for the environment variables.

I changed it slightly to request `DATABASE_URI` style for services. The general naming convention of avoiding short names (e.g. `DATABASE`, not `DB) still applies here.

Is there any other convention we follow and needs explicit documentation?

dennwc commented on engineering/conventions-go.md:
```diff
@@ -69,3 +69,9 @@ A blank line is usually added after the end of any control structure, too.
+## Configuration
+
+* Use environment variables for configuration.
+* You can use the [envconfig](https://github.com/kelseyhightower/envconfig) package to parse the environment into a configuration struct.
```
I think an exception could be made for some user-facing CLI applications like `bblfshctl`.

Image a case when you have a local and a remote `bblfshd` and you have to set 3 environment variables for each call (or open two terminals) to switch servers. It's easier to provide a config to the CLI.

Note that my point is not related to `bblfshd` itself, since it's easier to configure dockerized applications via environment variables.

As a middle ground, there is a project called [viper](https://github.com/spf13/viper) that allows setting CLI flags from env/config interchangeably.

smola replied:
Environment variables should be present in any case. This allows to set these either globally in your machine or in a given session, set them in a Kubernetes pod meant to be used interactively, etc.

We have discussed, and I think with no definite conclusion yet, about CLI flags being present in some cases _in addition to_ environment variables. This might be useful, specially if these are documented together in `--help` just as viper, go-flags and others do. In go-flags you can do it with the `env` struct tag.

mcuadros replied:
@dennwc  go-flags, supports envconfig too. Personally, I don't like at all viper or cobra.

@smola  Can you replace `You can use` with `You MUST use`?

dennwc replied:
@mcuadros Any specific concerns about cobra except that it's easier to use structs with field tags?

smola commented:
Merging, since there's no discussion with current content, just asking for additional info.
I will add further details once we discuss over concrete [borges example](https://github.com/src-d/borges/pull/294).
